daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/LayerTransformer.scala

```diff
@@ -200,6 +214,54 @@ abstract class LayerTransformer()
   //    println("DIS hex dump = " + str)
   //    ???
   //  }
+
+  /**
+   * Enables the layer transformer to take necessary state-maniuplating actions before
+   * the parsing of a layer starts.
+   *
+   * Examples might be setting variable values.
+   *
+   * At the time this is called, the state contains the layered I/O stream already.
+   * @param s The parser state, which may be modified.
+   */
+  def startLayerForParse(s: PState):Unit
+
+  /**
+   * Enables the layer transformer to take necessary state-maniuplating actions after
+   * the parsing of a layer ends.
+   *
+   * Examples might be setting variable values.
+   *
+   * At the time this is called, the state still contains the layered I/O stream.
+   * @param s The parser state, which may be modified.
+   */
+  def endLayerForParse(s: PState): Unit
+
+  /**
+   * Enables the layer transformer to take necessary state-maniuplating actions before
+   * the unparsing of a layer starts.
+   *
+   * Examples might be setting variable values.
+   *
+   * Note that if variables without values are read by this method, then this must create
+   * suspended calculations that clone the necessary state, which are evaluated later.
+   *
+   * At the time this is called, the state still contains the layered I/O stream.
+   * @param s The unparser state, which may be modified.
+   */
+  def startLayerForUnparse(s: UState): Unit
+
+  /**
+   * Enables the layer transformer to take necessary state-maniuplating actions after
+   * the unparsing of a layer ends.
+   *
+   * Note that if variables without values are read by this method, then this must create
+   * suspended calculations that clone the necessary state, which are evaluated later.
```

Comment:
This is a very good point. I think the plugin API really needs to abstract away this stuff so that a framework that creates suspensions is being used, and drives the plug in methods.